tests/equality/equality.cpp

```diff
@@ -0,0 +1,1060 @@
+    u2->setImportReference("some_location");
+
+    u1->setSourceUnits(is, "ampere");
+    u2->setSourceUnits(is, "ampere");
+
+    u1->addUnit("second", 1.0, "unit_id");
+    u2->addUnit("second", 1.0, "unit_id");
+
+    EXPECT_TRUE(u1->equals(u2));
+    EXPECT_TRUE(u2->equals(u1));
+}
+
+TEST(Equality, unitsNotEqualDifferingNumberOfUnits)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    libcellml::ImportSourcePtr is = libcellml::ImportSource::create();
+
+    u1->addUnit("second", 1.0, "second_id");
+    u1->addUnit("metre", -1.0, "metre_id");
+    u2->addUnit("second", 1.0, "second_id");
+
+    EXPECT_FALSE(u1->equals(u2));
+    EXPECT_FALSE(u2->equals(u1));
+}
+
+TEST(Equality, unitsNotEqualDifferingUnitsBasic)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    libcellml::ImportSourcePtr is = libcellml::ImportSource::create();
+
+    u1->addUnit("second", 1.0, "unit_id");
+    u2->addUnit("second", 2.0, "unit_id");
+
+    EXPECT_FALSE(u1->equals(u2));
+    EXPECT_FALSE(u2->equals(u1));
+}
+
+TEST(Equality, unitsEqualDifferingOrderOfUnits)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    u1->addUnit("second", 1.0, "unit_id");
+    u1->addUnit("metre", 3.0);
+    u2->addUnit("metre", 3.0);
+    u2->addUnit("second", 1.0, "unit_id");
+
+    EXPECT_TRUE(u1->equals(u2));
+    EXPECT_TRUE(u2->equals(u1));
+}
+
+TEST(Equality, unitsEqualDifferingOrderOfUnitsManyUnit)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    u1->addUnit("second", 1.0, "unit_id");
+    u1->addUnit("metre", 3.0);
+    u1->addUnit("apple", 1.0);
+    u1->addUnit("jedi", "yota", -2.0);
+    u1->addUnit("flip");
+    u1->addUnit("lumen");
+
+    u2->addUnit("jedi", "yota", -2.0);
+    u2->addUnit("second", 1.0, "unit_id");
+    u2->addUnit("lumen");
+    u2->addUnit("metre", 3.0);
+    u2->addUnit("apple", 1.0);
+    u2->addUnit("flip");
+
+    EXPECT_TRUE(u1->equals(u2));
+    EXPECT_TRUE(u2->equals(u1));
+}
+
+TEST(Equality, unitsNotEqualDifferingOrderOfUnits)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    u1->addUnit("metre", 3.0);
+    u1->addUnit("second", 1.0, "unit_id");
+    u2->addUnit("second", 2.0, "unit_id");
+    u2->addUnit("metre", 3.0);
+
+    EXPECT_FALSE(u1->equals(u2));
+    EXPECT_FALSE(u2->equals(u1));
+}
+
+TEST(Equality, unitsEqualPrecision1)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    u1->addUnit("second", 1.0);
+    u2->addUnit("second", 1.0 + std::numeric_limits<double>::epsilon());
```

Comment:
this confirms that the way we are comparing real numbers is behaving as expected, right? which is a good thing to test...

Reply:
I guess I was a bit taken aback by the numbers `238487722399.29383` and `238487722399.29384` to be considered nearly equal, even though they differ at the 5th decimal. However, I appreciate that those are big numbers, so a difference at the 5th decimal should indeed make those numbers nearly equal.

So, ok, I am fine with it. One thing, though, I think we should replace our existing `areEqual()` method in `utilities.cpp` with `nearlyEqual()` and, in fact, probably call that method `areNearlyEqual()`. In other words, I would remove `ulpsDistance()`, `nearlyEqual()` and `areEqual()`, and instead have:

```c++
bool areNearlyEqual(double a, double b)
{
    // Based on https://bitbashing.io/comparing-floats.html.

    // Classical equality check.

    static const auto Epsilon = std::numeric_limits<double>::epsilon();

    if (fabs(a - b) <= Epsilon) {
        return true;
    }

    // a and b must be of the same sign.

    if ((a < 0) != (b < 0)) {
        return false;
    }

    // The classical equality check wasn't successful, yet a and b are of the
    // same sign, so compute and check their ULPs distance.

    static const auto MaxUlpsDistance = std::numeric_limits<ptrdiff_t>::max();

    ptrdiff_t ulpsDistance = MaxUlpsDistance;

    if (!std::isnan(a) && !std::isnan(b) && !std::isinf(a) && !std::isinf(b)) {
        static const auto SizeOfDouble = sizeof(double);

        ptrdiff_t ia;
        ptrdiff_t ib;

        memcpy(&ia, &a, SizeOfDouble);
        memcpy(&ib, &b, SizeOfDouble);

        ulpsDistance = ia - ib;

        if (ulpsDistance < 0) {
            ulpsDistance = -ulpsDistance;
        }
    }

    static const ptrdiff_t UlpsEpsilon = 1;

    return ulpsDistance <= UlpsEpsilon;
}
```

Note that we don't check for `std::isinf(a) != std::isinf(b)` anymore. Indeed, it's not in the [original code](https://bitbashing.io/comparing-floats.html), so not sure how it ended up here, not least since there is no test for it.

Reply:
Because nothing is exact when it comes to real number operations.  Here we are saying that we are setting our tolerance for arithmetic error to be 1 ULPs. `unitsEqualPrecision3` and `unitsEqualPrecision4` say the same thing.  I guess welcome to the world of number representation in computers.

Reply:
Yes, I am now fine with those tests, but what about https://github.com/cellml/libcellml/pull/785#discussion_r585481555?

Reply:
`areEqual` and `areNearlyEqual` are different methods that serve different purposes.  So no they cannot be combined. `areNearlyEqual` is for numbers that (maybe) are the result of a mathematical operation.

Reply:
I cannot see why you would not break that function down into its constituent parts.

Reply:
There is a test for infinity, and there is a check for infinity in the original code.

Reply:
> `areEqual` and `areNearlyEqual` are different methods that serve different purposes. So no they cannot be combined. `areNearlyEqual` is for numbers that (maybe) are the result of a mathematical operation.

That's the thing, `areEqual()` and `areNearlyEqual()` actually serve the same purpose. `areEqual()` uses the classical equality check, which happens to be good enough for our current use of `areEqual()`. So, it's the same whether we use `areEqual()` or `areNearlyEqual()` for our current uses of `areEqual()`.

> I cannot see why you would not break that function down into its constituent parts.

Come on, it's a dead simple method. No need to break it down. It makes code maintenance unnecessarily "complicated".

Reply:
> There is a test for infinity, and there is a check for infinity in the original code.

I am clearly missing something since I can't see it...?!

Reply:
In equality.cpp line 339.

Reply:
> That's the thing, areEqual() and areNearlyEqual() actually serve the same purpose.

I disagree. `areEqual()` determines if in this case two doubles are exactly equal. `areNearlyEqual()` determines if two doubles are equal allowing for numerical error.

> Come on, it's a dead simple method. No need to break it down. It makes code maintenance unnecessarily "complicated".

I can't agree with that, calculating the ULPs distance is  a definite block of code that makes the function much more readable.  It in no way makes code maintenance unnecessarily complicated.

Reply:
> In equality.cpp line 339.

I guess you mean `TEST(Equality, unitsNotEqualInfinity)`? Anyway, I am talking about the original code at https://bitbashing.io/comparing-floats.html. I cannot see any `xor` operation there, just an `||` operation.



> > That's the thing, areEqual() and areNearlyEqual() actually serve the same purpose.
> I disagree. `areEqual()` determines if in this case two doubles are exactly equal. `areNearlyEqual()` determines if two doubles are equal allowing for numerical error.

`areEqual()` is currently defined as:
```c++
bool areEqual(double value1, double value2)
{
    return std::abs(value1 - value2) < std::numeric_limits<double>::epsilon();
}
```

which is the classical equality check in [my comment above](https://github.com/cellml/libcellml/pull/785#discussion_r585481555) (except for the fact that `areEqual()` uses `<` rather than `<=`).

> > Come on, it's a dead simple method. No need to break it down. It makes code maintenance unnecessarily "complicated".
> I can't agree with that, calculating the ULPs distance is a definite block of code that makes the function much more readable. It in no way makes code maintenance unnecessarily complicated.

It's a short function, called once, but never mind.